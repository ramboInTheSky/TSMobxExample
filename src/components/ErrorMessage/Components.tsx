import styled from '@emotion/styled'
import { variables } from '../../constants/theme'

export const Wrapper = styled.div`
	margin: auto;
	border: 5px solid #d43333;
	padding: 2rem;
    background-color: ${variables.componentBackgroundClear};
    margin-bottom: 2.5rem;
    color: #333;
    & h3{
        margin-top:0;
        color: #d43333;
    }
`
