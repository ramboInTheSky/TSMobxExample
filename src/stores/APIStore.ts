import { observable, action, toJS } from 'mobx'
import axios from 'axios'

import { getErrorCode, toSnakeCase, toCamelCase } from '../utils'

export { observable, action }

export class Store<T> {
	private url: string
	private fresh = false

	@observable protected data: Array<T> = []

	@observable protected detail: T | Partial<T> = {}

	@observable protected count: number = 0

	@observable protected loading: boolean = false

	@observable public error?: string = undefined

	constructor(uri: string) {
		this.url = uri
	}
	// public API
	// public getItems = (): Array<T> => {
	// 	return toJS(this.data) as Array<T>
	// }

	public get items(): Array<T> {
		return toJS(this.data) as Array<T>
	}

	public get item(): T {
		return toJS(this.detail) as T
	}

	// public getItem = (): T => {
	// 	return toJS(this.detail) as T
	// }

	public get errorMessage(): string {
		return this.error as string
	}

	public get isLoading(): boolean {
		return this.loading
	}

	public get isError(): boolean {
		return !!this.error
	}

	public isFresh = () => this.fresh
	// end public API

	protected apiEnded = (e?: any, customMessage?: string) => {
		this.loading = false
		this.error = e ? customMessage || getErrorCode(e) : undefined
		this.fresh = true
		const self = this
		setTimeout(function () {
			self.fresh = false
		})
	}

	@action
	async getList(url?: string, filter?: string) {
		if (!this.url && !url) return
		try {
			this.loading = true
			let res: any
			res = await axios.get(url || this.url + '?filter=' + filter)
			this.apiEnded()
			this.count = res.data && res.data.count ? res.data.count : 0
			return toCamelCase(res.data)
		} catch (e) {
			this.apiEnded(e)
			this.count = 0
			return []
		}
	}

	@action
	async getDetail(item: string) {
		try {
			this.loading = true
			let res: any
			res = await axios.get(`${this.url}/${item}`)
			this.apiEnded()
			return toCamelCase(res.data)
		} catch (e) {
			this.apiEnded(e)
			return {}
		}
	}

	@action
	async save(item: Partial<T>, id?: string, url?: string) {
		try {
			this.loading = true
			let res: any
			if (id) {
				res = await axios.put(`${url || this.url}/${id}`, toSnakeCase(item))
			} else {
				res = await axios.post(url || this.url, toSnakeCase(item))
			}
			this.apiEnded()
			return toCamelCase(res.data)
		} catch (e) {
			this.apiEnded(e)
			return {}
		}
	}
}
