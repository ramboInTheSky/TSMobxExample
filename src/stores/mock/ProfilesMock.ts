import { Profile } from '../../models/Profile'

export class ProfilesMock {
    public item: Profile = {
        id: '1aa657bc-d556-4643-a68f-af46f1bdba89',
        firstName: 'Clark',
        lastName: 'Kent',
        company: 'test 123',
        mobileNumber: '01234567890',
        emailAddress: 'dsg@dsf.com',
        notes: 'fdsdfsfs',
        created: '2019-04-16T12:03:37.493156',
        visits: [
            {
                id: 'd3aa88e2-c754-41e0-8ba6-4198a34aa0a2',
                lockGroups: ['Default', 'Boiler', '1.11'],
                fromDate: '2015-03-23T11:00:00Z',
                toDate: '2015-11-25T12:00:00Z',
                tags: ['123123', '234234', '345345'],
                archetypeIds: [],
            },
        ],
    }
    public detail = this.item
    public items: Profile[] = [
        {
            id: '1aa657bc-d556-4643-a68f-af46f1bdba89',
            firstName: 'Clark',
            lastName: 'Kent',
            company: 'test 123',
            mobileNumber: '01234567890',
            emailAddress: 'dsg@dsf.com',
            notes: 'fdsdfsfs',
            created: '2019-04-16T12:03:37.493156',
            visits: [
                {
                    id: 'd3aa88e2-c754-41e0-8ba6-4198a34aa0a2',
                    lockGroups: ['Default', 'Boiler', '1.11'],
                    fromDate: '2015-03-23T11:00:00Z',
                    toDate: '2015-11-25T12:00:00Z',
                    tags: ['123123', '234234', '345345'],
                    archetypeIds: [],
                },
            ],
        },
    ]
    public data = this.items
    public isLoading: boolean = false
    public isError: boolean = false
    public getDetail: Function = jest.fn()
    public getList: Function = jest.fn()
    public toggleEditing: Function = jest.fn()
}
