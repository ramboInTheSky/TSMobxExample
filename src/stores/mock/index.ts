export * from './RoutingMock'
export * from './ProfilesMock'
export * from './VisitsMock'

export const mockDateConstructor = () => {
    const now = new Date('2020')
    /*eslint no-global-assign:off*/
    Date = class extends Date {
        constructor() {
            super()
            return now
        }
    } as DateConstructor
}