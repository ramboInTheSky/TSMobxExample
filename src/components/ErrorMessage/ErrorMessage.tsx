import React from 'react'
import {Link} from 'react-router-dom'
import { observer, inject } from 'mobx-react'
import {Wrapper} from './Components'
import { StoreNames, Stores, Routing } from '../../stores';

export interface ErrorMessageProps {
	children?: string
}
export interface ErrorMessageConnectedProps extends ErrorMessageProps {
	routing: Routing
}
@inject(StoreNames.routing)
@observer
export class ErrorMessage extends React.Component<ErrorMessageProps> {
	get store(){
		return this.props as ErrorMessageConnectedProps
	}
	render(){
		const onClick = ()=> {
			this.store.routing.goToPage('/')
		}
	return <Wrapper><h3>Something has gone wrong!</h3>{this.props.children}<br /><br /><Link to="/" onClick={onClick}>Go back to home page</Link></Wrapper>
	}
}
