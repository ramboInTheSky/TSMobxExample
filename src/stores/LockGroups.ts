import { observable, action } from 'mobx'
import { Store } from './APIStore'
import { LockGroup } from '../models/LockGroups'
import urls from '../constants/urls'

const sortingFn = (a: LockGroup, b: LockGroup) => {
	const first = a.name
	const last = b.name
	if (first > last) return -1
	if (last > first) return 1
	return 0
}

export class LockGroupStore extends Store<LockGroup> {
	@observable
	public selectedLockGroups: Array<LockGroup> = []

	constructor() {
		super(urls.lockGroups)
	}

	async getList(filter?: string) {
		const data = await super.getList(undefined, filter)
		this.data =
			data && data.items ? data.items : []
	}

	@action 
	setList(newList: Array<LockGroup>){
		this.data = newList
	}

}
