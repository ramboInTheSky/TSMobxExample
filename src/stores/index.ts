// first party
import { ProfileStore } from './Profiles'
import { Routing } from './Routing'
import { VisitStore } from './Visits'
import { LockGroupStore } from './LockGroups'

// re-exports
export * from './Profiles'
export * from './Routing'
export * from './Visits'
export * from './LockGroups'

export interface Stores {
	profiles: ProfileStore
	routing: Routing
	visits: VisitStore
	lockGroups: LockGroupStore
}

type StoreNamesGuard = keyof Stores

export class StoreNames {
	static profiles: StoreNamesGuard = 'profiles'
	static routing: StoreNamesGuard = 'routing'
	static visits: StoreNamesGuard = 'visits'
	static lockGroups: StoreNamesGuard = 'lockGroups'
}

const localStores: Partial<Stores> = {
	profiles: new ProfileStore(),
	routing: new Routing(),
	visits: new VisitStore(),
	lockGroups: new LockGroupStore(),
}

//test comment
export const stores = { ...localStores }
