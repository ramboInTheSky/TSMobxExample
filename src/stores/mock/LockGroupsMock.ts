import { LockGroup } from '../../models/LockGroups';

export class LockGroupMock {
    public detail: LockGroup = new LockGroup()
    public data = [
        {
            "id": "123-123-123-123",
            "name": "1.11",
            "type": "room 11",
            "notes": "this is just a room"
        },
        {
            "id": "123-123-123-123",
            "name": "1.12",
            "type": "room 12",
            "notes": "this is the best room"
        },
        {
            "id": "123-123-123-123",
            "name": "1.13",
            "type": "room 13",
            "notes": "this is just a room"
        },
        {
            "id": "123-123-123-123",
            "name": "1.14",
            "type": "room 14",
            "notes": "this is just a room"
        },
        {
            "id": "123-123-123-123",
            "name": "1.15",
            "type": "room 15",
            "notes": "this is just a room"
        },
        {
            "id": "123-123-123-123",
            "name": "1.16",
            "type": "room 16",
            "notes": "this is just a room"
        },
        {
            "id": "123-123-123-123",
            "name": "1.17",
            "type": "room 17",
            "notes": "this is an unlucky room"
        }
    ]
    public getList = jest.fn()
    public items = []
    public item = () => this.detail
}