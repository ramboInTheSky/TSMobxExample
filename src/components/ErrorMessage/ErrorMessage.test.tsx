import React from 'react'
import { shallow } from 'enzyme'

import { ErrorMessage,  ErrorMessageProps} from './ErrorMessage'
import { RoutingMock } from '../../stores/mock';

describe('<ErrorMessage />', () => {
	let props: any
	beforeEach(()=>{
		props = {
			routing: new RoutingMock()
		}
	})
	it('renders correctly', () => {
		expect(shallow(<ErrorMessage {...props}/>)).toMatchSnapshot()
	})

	// it('passes on props correctly', () => {
	// 	expect(shallow(<ErrorMessage foo="bar">Hello</ErrorMessage>)).toMatchSnapshot()
	// })

	// it('Does not re-render on children props change', () => {
	// 	const el: any = shallow(<ErrorMessage foo="bar" />)
	// 	el.setProps({ children: 'Goodbye' })
	// 	el.update()
	// 	expect(el).toMatchSnapshot()
	// })
})
