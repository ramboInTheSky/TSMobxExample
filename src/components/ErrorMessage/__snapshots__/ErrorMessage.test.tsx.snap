// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorMessage /> renders correctly 1`] = `
<ErrorMessage
  routing={
    RoutingMock {
      "breadcrumb": Array [],
      "getBreadcrumb": [MockFunction],
      "go": [MockFunction],
      "goBack": [MockFunction],
      "goBackToPage": [MockFunction],
      "goForward": [MockFunction],
      "goToPage": [MockFunction],
      "history": Object {},
      "location": Object {},
      "push": [MockFunction],
      "replace": [MockFunction],
      "splitThings": [MockFunction],
    }
  }
/>
`;
